### ISSUE TITLE
Add span properties to slices in C#?

### ISSUE THREAD

**Issue by Skippeh** [ralfbiedert/interoptopus#35]
I feel like there's no good way to access all of (instead of indexer (which copies data anyway)) the slice data in c# right now without copying it.

I propose adding a method to the c# generated version of FFISlice and FFISliceMut that has a ReadOnlySpan<T> and Span<T> respectively.

For example:
```csharp
///A pointer to an array of data someone else owns which may be modified.
[Serializable]
[StructLayout(LayoutKind.Sequential)]
public partial struct SliceMutf32
{
    ///Pointer to start of mutable data.
    IntPtr data;
    ///Number of elements.
    ulong len;
}

public partial struct SliceMutf32
{
    public Span<float> AsSpan()
    {
        unsafe
        {
            return new Span<float>((float*)data, (int)len);
        }
    }
}
```

Not sure how you'd do it without unsafe though.

Should probably also make it optional, since Span is part of System.Memory which isn't included in the standard framework on all .net versions.

**Comment by ralfbiedert** [comment]
I'm not opposed to using `Span`, but I don't get this:

> there's no good way to access all of (instead of indexer (which copies data anyway)) the slice data in c# right now without copying it.

Once you have a `Slice...`, doesn't this give you access without copying it? This code is generated already when enabling C# `unsafe`.

```csharp

        public uint this[int i]
        {
            get
            {
                if (i >= Count) throw new IndexOutOfRangeException();
                unsafe
                {
                    var d = (uint*) data.ToPointer();
                    return d[i];
                }
            }
            set
            {
                if (i >= Count) throw new IndexOutOfRangeException();
                unsafe
                {
                    var d = (uint*) data.ToPointer();
                    d[i] = value;
                }
            }
        }

```

**Comment by Skippeh** [comment]
Not if you want to pass the entire slice to something else in C#.

One solution would be to pass ```Slice...``` itself i guess, but in my case it's internal so it's not visible to other assemblies.

**Comment by Skippeh** [comment]
Thanks for the fix! Method is indeed probably the way to go since it returns a new span everytime you call it.

**Comment by ralfbiedert** [comment]
Hm, I agree that "same == property", "differing == method" logic makes sense, but shouldn't it be a property then since the returned `Span` is always the same? I mean, it's a new instance, but a struct at that with the same values.

...

Ok, [I just read up more](https://docs.microsoft.com/en-us/previous-versions/dotnet/netframework-1.1/bzwdh01d(v=vs.71)?redirectedfrom=MSDN#cpconpropertyusageguidelinesanchor1
) on the matter. According to that link:

> Use a method when
> - The operation is a conversion, such as Object.ToString.

Yes, it is a conversion; +1 for method.

> - The operation is expensive enough that you want to communicate to the user that they should consider caching the result.

No, not expensive: +1 property.

> - Obtaining a property value using the get accessor would have an observable side effect.

No side effects, +1 property.

> - Calling the member twice in succession produces different results.

No differing results, +1 property.

> - The order of execution is important. Note that a type's properties should be able to be set and retrieved in any order.

No order impact, +1 property.

> - The member is static but returns a value that can be changed.

N/A

> - The member returns an array. Properties that return arrays can be very misleading. Usually it is necessary to return a copy of the internal array so that the user cannot change internal state. This, coupled with the fact that a user can easily assume it is an indexed property, leads to inefficient code. In the following code example, each call to the Methods property creates a copy of the array. As a result, 2n+1 copies of the array will be created in the following loop.

It returns something similar to an array, but, the whole point of Spans is that they reflect (and mutate) the internal state, +1 property.

All things considered property seems to be more in line with the guide. Since that's what's in 0.14 already I'll stay with it unless you / someone screams.

**Comment by Skippeh** [comment]
Yeah when you put it like that properties seem like the way to go. 👍
